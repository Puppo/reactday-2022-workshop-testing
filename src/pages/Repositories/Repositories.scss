.Repositories {
  text-align: center;

  &-results {
    margin: 30px 20px 0;
  }
}
