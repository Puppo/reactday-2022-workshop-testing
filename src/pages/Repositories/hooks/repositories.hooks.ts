import { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import {
  getRepositoriesByUser,
  Repository,
} from "../../../services/github.service";

export function useRepositories() {
  const { userId: username } = useParams();

  const [repositories, setRepositories] = useState<Repository[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | undefined>();

  useEffect(() => {
    async function load() {
      if (!username) return;

      setLoading(true);
      setError(undefined);
      try {
        const repositories = await getRepositoriesByUser(username);
        setRepositories(repositories);
      } catch (error) {
        let message = "An error occurred while loading repositories";
        if (error instanceof Error) message = error.message;
        setError(message);
      }
      setLoading(false);
    }
    load();
  }, [username]);

  return {
    repositories,
    loading,
    error,
  };
}
