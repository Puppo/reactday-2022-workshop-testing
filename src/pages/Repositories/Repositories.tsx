import { Loader, Notification, RepositoryCard } from '../../components';
import { arrayUtils } from '../../utils';
import { useRepositories } from './hooks/repositories.hooks';
import './Repositories.scss';

export default function Repositories() {

  const { repositories, error, loading } = useRepositories();
  const repositoriesChunks = arrayUtils.sliceIntoChunks(repositories, 4);

  return (
    <div className='Repositories'>
      <h1 className='title'>Repositories</h1>

      {loading && <Loader />}
      {error && <Notification type="error" text={error} />}
      {!loading && !error && <div className="Repositories-results">
        {repositoriesChunks.map((repositories, idx) => <div className="columns" key={idx}>
          {repositories.map(({ name, description }) => <div className="column is-3" key={name}>
            <RepositoryCard name={name} description={description} />
          </div>)}
        </div>)}
      </div>}
    </div>
  )
}
