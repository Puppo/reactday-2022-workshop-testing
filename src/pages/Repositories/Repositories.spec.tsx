import { render, screen } from "@testing-library/react";
import { Repository } from "../../services/github.service";
import * as useRepositoriesModule from "./hooks/repositories.hooks";
import Repositories from "./Repositories";


const repositoriesMock: Repository[] = [{
  name: 'test-name',
  description: 'test-description'
}, {
  name: 'test-name-2',
  description: 'test-description-2'
}];


describe('Repositories', () => {

  const mockUseRepositories = (data: ReturnType<typeof useRepositoriesModule['useRepositories']>) => {
    return jest.spyOn(useRepositoriesModule, 'useRepositories').mockReturnValueOnce(data);
  }

  it('should render the component', async () => {
    const spyUseRepositories = mockUseRepositories({ loading: false, error: undefined, repositories: repositoriesMock });

    const { container } = render(<Repositories />);
    expect(container).toMatchSnapshot();
    expect(spyUseRepositories).toHaveBeenCalled();
  });

  it('should render the error', async () => {
    mockUseRepositories({ loading: false, error: 'Error', repositories: [] });

    render(<Repositories />);

    expect(screen.getByText("Error")).toBeInTheDocument();
  });

  it('should render loader', async () => {
    mockUseRepositories({ loading: true, error: undefined, repositories: [] });

    render(<Repositories />);

    const loader = screen.queryByTitle(/content loading/i)
    expect(loader).toBeInTheDocument();
  });
})


export { };
