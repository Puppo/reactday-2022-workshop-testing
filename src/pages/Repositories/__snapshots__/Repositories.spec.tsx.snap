// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repositories should render the component 1`] = `
<div>
  <div
    class="Repositories"
  >
    <h1
      class="title"
    >
      Repositories
    </h1>
    <div
      class="Repositories-results"
    >
      <div
        class="columns"
      >
        <div
          class="column is-3"
        >
          <div
            class="card"
          >
            <div
              class="card-content"
            >
              <div
                class="media"
              >
                <div
                  class="media-content"
                >
                  <p
                    class="title is-4"
                  >
                    test-name
                  </p>
                </div>
              </div>
              <div
                class="content"
              >
                test-description
              </div>
            </div>
          </div>
        </div>
        <div
          class="column is-3"
        >
          <div
            class="card"
          >
            <div
              class="card-content"
            >
              <div
                class="media"
              >
                <div
                  class="media-content"
                >
                  <p
                    class="title is-4"
                  >
                    test-name-2
                  </p>
                </div>
              </div>
              <div
                class="content"
              >
                test-description-2
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
